### DISCUSSION TITLE
Paste range feature

### DISCUSSION THREAD

**Discussion by MartinDawson** [handsontable/hyperformula#763] [Feature requests]
Hi, sorry to post on this discussion board so much, but I think it's worth raising because currently I have no way around this issue. It's probably because I'm using every single feature of hyperformula..

So currently my spreadsheet has a paste range feature like Google Sheets & Excel.

One example of how it works is like this:

- User copies a single.
- User drags their range selection down and does Ctrl + V (pastes it).

The one cell is duplicated into the selected cells.

Here's a gif showing you what I mean from GSheets:

![gif7](https://user-images.githubusercontent.com/15030491/125354282-894d3680-e35b-11eb-8180-2d0435cddb3c.gif)

This is currently not possible in Hyperformula.

Here's the issue in my sheet which uses hyperformula:

![gif6](https://user-images.githubusercontent.com/15030491/125353474-84d44e00-e35a-11eb-8d98-f8f842f54d40.gif)

The only way to do it is to loop through the destination cells and use `hyperformula.paste(cellAddress)` one by one.

However this messes up the internal history, so when I do CTRL + Z (Undo) it only undos the pastes individually, one by one.

What should happen is that it should undo the entire range all at once.

So a paste range feature would solve this or a way for me to turn off and on the internal history of hyperformula so that I can implement the paste range for my styles and cells correctly and just make hyperformula undo remove it all at once.

Thank you

**Comment by wojciechczerniak** [comment]
Batch would create a single Undo step, but Clipboard does not work with batching. So no, for multiple `paste` calls. You can try `develop` that has a new method `getFillRangeData` that will get you calculated content. Sorry for source code link, but docs for 1.0 are not published yet:
https://github.com/handsontable/hyperformula/blob/71677a49af48e9e9b9ee2d12054e05341a086f7d/src/HyperFormula.ts#L2395-L2417


A sample implementation could look like:
```ts
onPaste(source: SimpleCellRange, target: SimpleCellRange) {
  if (hf.isItPossibleToSetCellContents(target)) {
    const data = hf.getFillRangeData(source, target)
    hf.setCellContents(target.start, data)
  }
}
```

**Reply by MartinDawson** [reply]
@wojciechczerniak That works perfect. thanks a lot!

It correctly handles formula references and undo/redo history is good

**Reply by MartinDawson** [reply]
Just saw a slight issue with it. I'm using the example exactly as provided above.

It seems like this is the expected behavior of `getFillRangeData` though so I'm not sure if we need another parameter or function.

When doing a copy/paste using it it's reversing the order of the copied cells if I'm on a row that isn't on the same nth number as the start of the copy selection.

It seems to be this line that returns the data in a different order than I was expecting:
 `const data = this.hyperformula.getFillRangeData(source, target);`

Here's a visual gif: 

![gif7](https://user-images.githubusercontent.com/15030491/125486781-41f0a69d-9a3a-497d-9dce-ac0cd5d13e82.gif)

So you can see it correctly copies and fills out the cells with wrap-around heuristics but I want to keep the top destination cells to be the same order as the top of the source cells and only wrap the bottom ones, i.e like Google Sheets does.

Because if a user is copy/pasting a range they don't want the order of their copy to be swapped around at the top level. It only should wrap at the bottom of it.

So if I do the copy/paste on the row below it works fine because it's on the same nth level as 54 is:

![gif8](https://user-images.githubusercontent.com/15030491/125487192-2a509658-cecd-4644-a39e-87444dcbb17c.gif)

Here you can see that Google Sheets does wrap-around heuristics but **not** for the top level for the copy/paste, i.e the 2222 stays on the top level of the pasted values..

![gif8](https://user-images.githubusercontent.com/15030491/125487736-be760d4b-35ce-4524-bc6d-ac61d3a6883a.gif)

So maybe we could do with another parameter to this function that allows us to not wrap the top level?

Thanks

**Reply by MartinDawson** [reply]
https://github.com/handsontable/hyperformula/pull/767 has fixed this issue.

Thanks

**Reply by wojciechczerniak** [reply]
Oh right, I forgot to reply what we're doing. As you already found out, we've implemented your suggestion. A very good idea @MartinDawson, thanks

**Reply by MartinDawson** [reply]
No problem, thanks for the fast fixes, really amazing speed and quality by the maintainers here

**Comment by MartinDawson** [comment]
I just ran into an issue with this @wojciechczerniak:

The docs say this: `* @throws [[SheetsNotEqual]] if both ranges are not from the same sheet`

So it does not work across different sheets unfortunately which basically means our copy/paste doesn't work across multiple sheets.

Is this just a current limitation with plans to allow it cross-sheets or is it not possible?

**Reply by wojciechczerniak** [reply]
True. It was created with a specific use case in mind. There are no plans to make it work between sheets, as no one yet mentioned such a use case. Feel free to report a feature request in issues.

@izulin Is it possible to alter the algorithm? ~Maybe if we get two different sheets we could copy the content first and then apply the same algorithm to fill the range? Preferably in `batch` to have a single `undo` action~ (as @izulin noted, we do not modify data with this method so it can't do that)

**Reply by wojciechczerniak** [reply]
@MartinDawson We've removed same sheet limitation in #836
